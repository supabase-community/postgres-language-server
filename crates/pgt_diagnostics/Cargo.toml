[package]
authors.workspace    = true
categories.workspace = true
description          = "<DESCRIPTION>"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "pgt_diagnostics"
repository.workspace = true
version              = "0.0.0"


[dependencies]
backtrace                  = "0.3.74"
bpaf                       = { workspace = true }
enumflags2                 = { workspace = true }
oxc_resolver               = { workspace = true }
pgt_console                = { workspace = true, features = ["serde"] }
pgt_diagnostics_categories = { workspace = true, features = ["serde"] }
pgt_diagnostics_macros     = { workspace = true }
pgt_text_edit              = { workspace = true, features = ["serde"] }
pgt_text_size.workspace    = true
schemars                   = { workspace = true, optional = true }
serde                      = { workspace = true, features = ["derive"] }
serde_json                 = { workspace = true }
termcolor                  = { workspace = true }
unicode-width              = { workspace = true }

[features]
schema = ["dep:schemars", "pgt_text_edit/schema", "pgt_diagnostics_categories/schema", "pgt_console/schema"]

[dev-dependencies]

[lib]
doctest = false
