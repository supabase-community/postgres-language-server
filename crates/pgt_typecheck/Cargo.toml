[package]
authors.workspace    = true
categories.workspace = true
description          = "<DESCRIPTION>"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "pgt_typecheck"
repository.workspace = true
version              = "0.0.0"


[dependencies]
pgt_console.workspace            = true
pgt_diagnostics.workspace        = true
pgt_query_ext.workspace          = true
pgt_schema_cache.workspace       = true
pgt_text_size.workspace          = true
pgt_treesitter.workspace = true
sqlx.workspace                   = true
tokio.workspace                  = true
tree-sitter.workspace            = true
tree_sitter_sql.workspace        = true

[dev-dependencies]
insta.workspace          = true
pgt_test_utils.workspace = true

[lib]
doctest = false
