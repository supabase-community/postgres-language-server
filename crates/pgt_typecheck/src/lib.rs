mod diagnostics;
mod typed_identifier;

pub use diagnostics::TypecheckDiagnostic;
use diagnostics::create_type_error;
use globset::Glob;
use itertools::Itertools;
use pgt_schema_cache::SchemaCache;
use sqlx::postgres::PgDatabaseError;
pub use sqlx::postgres::PgSeverity;
use sqlx::{Executor, PgPool};
use typed_identifier::apply_identifiers;
pub use typed_identifier::{IdentifierType, TypedIdentifier};

#[derive(Debug)]
pub struct TypecheckParams<'a> {
    pub conn: &'a PgPool,
    pub sql: &'a str,
    pub ast: &'a pgt_query::NodeEnum,
    pub tree: &'a tree_sitter::Tree,
    pub schema_cache: &'a pgt_schema_cache::SchemaCache,
    pub identifiers: Vec<TypedIdentifier>,
    /// Set of glob patterns that will be matched against the schemas in the database.
    /// Each matching schema will be added to the search_path for the typecheck.
    pub search_path_patterns: Vec<String>,
}

pub async fn check_sql(
    params: TypecheckParams<'_>,
) -> Result<Option<TypecheckDiagnostic>, sqlx::Error> {
    // Check if the AST is not a supported statement type
    if !matches!(
        params.ast,
        pgt_query::NodeEnum::SelectStmt(_)
            | pgt_query::NodeEnum::InsertStmt(_)
            | pgt_query::NodeEnum::UpdateStmt(_)
            | pgt_query::NodeEnum::DeleteStmt(_)
            | pgt_query::NodeEnum::CommonTableExpr(_)
    ) {
        return Ok(None);
    }

    let mut conn = params.conn.acquire().await?;

    // Postgres caches prepared statements within the current DB session (connection).
    // This can cause issues if the underlying table schema changes while statements
    // are cached. By closing the connection after use, we ensure a fresh state for
    // each typecheck operation.
    conn.close_on_drop();

    let (prepared, range_adjustments) = apply_identifiers(
        params.identifiers,
        params.schema_cache,
        params.tree,
        params.sql,
    );

    let mut search_path_schemas =
        get_schemas_in_search_path(params.schema_cache, params.search_path_patterns);

    if !search_path_schemas.is_empty() {
        // Always include public if we have any schemas in search path
        if !search_path_schemas.contains(&"public") {
            search_path_schemas.push("public");
        }

        let search_path_query = format!("SET search_path TO {};", search_path_schemas.join(", "));
        conn.execute(&*search_path_query).await?;
    }

    let res = conn.prepare(&prepared).await;

    match res {
        Ok(_) => Ok(None),
        Err(sqlx::Error::Database(err)) => {
            let pg_err = err.downcast_ref::<PgDatabaseError>();
            Ok(Some(create_type_error(
                pg_err,
                params.tree,
                range_adjustments,
            )))
        }
        Err(err) => Err(err),
    }
}

fn get_schemas_in_search_path(schema_cache: &SchemaCache, glob_patterns: Vec<String>) -> Vec<&str> {
    // iterate over glob_patterns on the outside to keep the order
    glob_patterns
        .iter()
        .filter_map(|pattern| {
            if let Ok(glob) = Glob::new(pattern) {
                let matcher = glob.compile_matcher();

                Some(
                    schema_cache
                        .schemas
                        .iter()
                        .filter_map(|s| {
                            if matcher.is_match(s.name.as_str()) {
                                Some(s.name.as_str())
                            } else {
                                None
                            }
                        })
                        .collect::<Vec<&str>>(),
                )
            } else {
                None
            }
        })
        .flatten()
        .unique()
        .collect()
}
