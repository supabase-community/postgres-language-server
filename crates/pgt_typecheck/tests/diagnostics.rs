use pgt_console::fmt::{Formatter, HTML};
use pgt_diagnostics::Diagnostic;
use pgt_typecheck::{IdentifierType, TypecheckParams, TypedIdentifier, check_sql};
use sqlx::{Executor, PgPool};
use std::fmt::Write;

struct TestSetup<'a> {
    name: &'a str,
    query: &'a str,
    setup: Option<&'a str>,
    test_db: &'a PgPool,
    typed_identifiers: Vec<TypedIdentifier>,
}

impl TestSetup<'_> {
    async fn test(self) {
        if let Some(setup) = self.setup {
            self.test_db
                .execute(setup)
                .await
                .expect("Failed to setup test selfbase");
        }

        let mut parser = tree_sitter::Parser::new();
        parser
            .set_language(&pgt_treesitter_grammar::LANGUAGE.into())
            .expect("Error loading sql language");

        let schema_cache = pgt_schema_cache::SchemaCache::load(self.test_db)
            .await
            .expect("Failed to load Schema Cache");

        let root = pgt_query::parse(self.query)
            .unwrap()
            .into_root()
            .expect("Failed to parse query");
        let tree = parser.parse(self.query, None).unwrap();

        let result = check_sql(TypecheckParams {
            conn: self.test_db,
            sql: self.query,
            ast: &root,
            tree: &tree,
            schema_cache: &schema_cache,
            identifiers: self.typed_identifiers,
            search_path_patterns: vec![],
        })
        .await;

        assert!(
            result.is_ok(),
            "Got Typechecking error: {}",
            result.unwrap_err()
        );

        let maybe_diagnostic = result.unwrap();

        let content = match maybe_diagnostic {
            Some(d) => {
                let mut result = String::new();

                if let Some(span) = d.location().span {
                    for (idx, c) in self.query.char_indices() {
                        if pgt_text_size::TextSize::new(idx.try_into().unwrap()) == span.start() {
                            result.push_str("~~~");
                        }
                        if pgt_text_size::TextSize::new(idx.try_into().unwrap()) == span.end() {
                            result.push_str("~~~");
                        }
                        result.push(c);
                    }
                } else {
                    result.push_str("~~~");
                    result.push_str(self.query);
                    result.push_str("~~~");
                }

                writeln!(&mut result).unwrap();
                writeln!(&mut result).unwrap();

                let mut msg_content = vec![];
                let mut writer = HTML::new(&mut msg_content);
                let mut formatter = Formatter::new(&mut writer);
                d.message(&mut formatter).unwrap();

                result.push_str(String::from_utf8(msg_content).unwrap().as_str());

                result
            }
            None => String::from("No Diagnostic"),
        };

        insta::with_settings!({
            prepend_module_to_snapshot => false,
        }, {
            insta::assert_snapshot!(self.name, content);

        });
    }
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn invalid_column(test_db: PgPool) {
    TestSetup {
        name: "invalid_column",
        query: "select id, unknown from contacts;",
        setup: Some(
            r#"
        create table public.contacts (
            id serial primary key,
            name varchar(255) not null,
            is_vegetarian bool default false,
            middle_name varchar(255)
        );
    "#,
        ),
        test_db: &test_db,
        typed_identifiers: vec![],
    }
    .test()
    .await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn invalid_type_in_function(test_db: PgPool) {
    // create or replace function clean_up(uid uuid)
    // returns void
    // language sql
    // as $$
    //     delete from public.contacts where id = uid;
    // $$;

    let setup = r#"
        create table public.contacts (
            id serial primary key,
            name text not null,
            is_vegetarian bool default false,
            middle_name varchar(255)
        );
    "#;

    /* NOTE: The replaced type default value is *longer* than the param name. */
    TestSetup {
        name: "invalid_type_in_function_longer_default",
        setup: Some(setup),
        query: r#"delete from public.contacts where id = uid;"#,
        test_db: &test_db,
        typed_identifiers: vec![TypedIdentifier {
            path: "clean_up".to_string(),
            name: Some("uid".to_string()),
            type_: IdentifierType {
                schema: None,
                name: "uuid".to_string(),
                is_array: false,
            },
        }],
    }
    .test()
    .await;

    /* NOTE: The replaced type default value is *shorter* than the param name. */
    TestSetup {
        name: "invalid_type_in_function_shorter_default",
        setup: None,
        query: r#"delete from public.contacts where id = contact_name;"#,
        test_db: &test_db,
        typed_identifiers: vec![TypedIdentifier {
            path: "clean_up".to_string(),
            name: Some("contact_name".to_string()),
            type_: IdentifierType {
                schema: None,
                name: "text".to_string(),
                is_array: false,
            },
        }],
    }
    .test()
    .await;
}
