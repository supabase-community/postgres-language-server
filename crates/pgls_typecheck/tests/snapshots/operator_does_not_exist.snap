---
source: crates/pgt_typecheck/tests/diagnostics.rs
expression: content
---
select ~~~is_active + product_name~~~ from public.products;

operator does not exist: boolean + unknown
